sparse/_coo/core.py

```diff
@@ -1485,10 +1488,10 @@ def pad(self, pad_width, mode="constant", **kwargs):
         if mode.lower() != "constant":
             raise NotImplementedError("Mode '{}' is not supported.".format(mode))
 
-        if kwargs.pop("constant_values", self.fill_value) != self.fill_value:
+        if kwargs.pop("constant_values", _zero_of_dtype(self.dtype)) != self.fill_value:
```

Comment:
You missed checking the `constant_values` with `equivalent`. If the fill-value is `nan` this will have issues, as `np.nan != np.nan` returns `True`.